File: src/cmd/outdated.zig
const std = @import("std");
const tool_mod = @import("../tool.zig");
const state_mod = @import("../state.zig");
const output = @import("../ui/output.zig");

const help =
    \\Usage: dot outdated
    \\
    \\List installed tools that have a newer version available.
    \\Resolves latest versions from the registry (requires network).
    \\
    \\Options:
    \\  --help, -h    Show this help
    \\
    \\Examples:
    \\  dot outdated
    \\  dot upgrade            # upgrade all outdated tools
    \\
;

pub fn run(
    allocator: std.mem.Allocator,
    args: []const []const u8,
    state: *state_mod.State,
    tools: []const tool_mod.Tool,
) !void {
    for (args) |a| {
        if (std.mem.eql(u8, a, "--help") or std.mem.eql(u8, a, "-h")) {
            output.printRaw(help);
            return;
        }
    }

    const OutdatedEntry = struct {
        name: []const u8,
        installed: []const u8,
        latest: []u8,
        pinned: bool,
    };

    var outdated: std.ArrayList(OutdatedEntry) = .empty;
    defer {
        for (outdated.items) |e| allocator.free(e.latest);
        outdated.deinit(allocator);
    }

    var checked: usize = 0;
    for (tools) |t| {
        const entry = state.tools.get(t.id) orelse continue;
        if (entry.version.len == 0) continue;
        checked += 1;

        const latest = t.version_source.resolve(allocator) catch continue;
        if (std.mem.eql(u8, latest, entry.version)) {
            allocator.free(latest);
            continue;
        }

        try outdated.append(allocator, .{
            .name = t.name,
            .installed = entry.version,
            .latest = latest,
            .pinned = entry.pinned,
        });
    }

    if (outdated.items.len == 0) {
        output.printSectionHeaderFmt("All {d} installed tool{s} are up to date.", .{
            checked, if (checked == 1) @as([]const u8, "") else "s",
        });
        output.printFmt("\n", .{});
        return;
    }

    output.printSectionHeaderFmt("{d} tool{s} have updates available", .{
        outdated.items.len, if (outdated.items.len == 1) @as([]const u8, "") else "s",
    });

    output.printFmt("\n{s}{s:<18} {s:<14} {s:<14} {s}{s}\n", .{
        output.bold, "Tool", "Current", "Latest", "Pinned", output.reset,
    });

    for (outdated.items) |e| {
        const name_trunc = e.name[0..@min(e.name.len, 17)];
        const cur_trunc = e.installed[0..@min(e.installed.len, 13)];
        const lat_trunc = e.latest[0..@min(e.latest.len, 13)];
        if (e.pinned) {
            output.printFmt("{s:<18} {s:<14} {s:<14} {s}{s}{s}\n", .{
                name_trunc, cur_trunc, lat_trunc, output.dim, output.sym_ok, output.reset,
            });
        } else {
            output.printFmt("{s:<18} {s:<14} {s:<14}\n", .{ name_trunc, cur_trunc, lat_trunc });
        }
    }

    output.printFmt("\nRun '{s}dot upgrade{s}' to upgrade all (pinned skipped unless --force).\n\n", .{
        output.bold, output.reset,
    });
}

File: src/cmd/outdated.rs
use crate::platform::PackageManager;
use crate::state::State;
use crate::tool::{InstallStrategy, Tool};
use crate::ui::output;
use clap::Args;
use console::style;

/// For system_package tools, compare against the package manager's own candidate
/// version instead of the upstream GitHub tag - the two are unrelated version
/// schemes, and the tag is often well ahead of what the distro repo can install.
/// Uses whichever variant (see `Tool::variants`) the tool was actually installed
/// under, so this matches what `dot upgrade` would actually do.
fn resolve_latest(tool: &Tool, state: &State) -> Option<String> {
    let strategy = match state.get_variant(&tool.id).and_then(|n| tool.variants.get(n)) {
        Some(v) => &v.strategy,
        None => &tool.strategy,
    };
    if let InstallStrategy::SystemPackage(s) = strategy {
        let pm = PackageManager::detect();
        if let Some(pkg) = s.package_for(pm) {
            if let Some(v) = pm.candidate_version(pkg) {
                return Some(v);
            }
        }
    }
    tool.version_source.resolve().ok()
}

#[derive(Debug, Args)]
pub struct OutdatedArgs {}

pub fn run(_args: &OutdatedArgs, state: &State, tools: &[Tool]) -> anyhow::Result<()> {
    struct OutdatedEntry<'a> {
        name: &'a str,
        installed: &'a str,
        latest: String,
        pinned: bool,
    }

    let mut outdated: Vec<OutdatedEntry> = Vec::new();
    let mut checked = 0usize;

    for t in tools {
        let entry = match state.get_entry(&t.id) {
            Some(e) if !e.version.is_empty() => e,
            _ => continue,
        };
        checked += 1;

        let latest = match resolve_latest(t, state) {
            Some(v) => v,
            None => continue,
        };

        if latest == entry.version {
            continue;
        }

        outdated.push(OutdatedEntry {
            name: &t.name,
            installed: &entry.version,
            latest,
            pinned: entry.pinned,
        });
    }

    if outdated.is_empty() {
        let s = if checked == 1 { "" } else { "s" };
        output::print_section_header(&format!("All {checked} installed tool{s} are up to date."));
        eprintln!();
        return Ok(());
    }

    let s = if outdated.len() == 1 { "" } else { "s" };
    output::print_section_header(&format!(
        "{} tool{s} have updates available",
        outdated.len()
    ));

    let colored = output::get_render_mode() == output::RenderMode::Rich;
    if colored {
        eprintln!(
            "\n{} {} {} {}",
            output::pad_to(&style("Tool").bold().to_string(), 18),
            output::pad_to(&style("Current").bold().to_string(), 14),
            output::pad_to(&style("Latest").bold().to_string(), 14),
            style("Pinned").bold(),
        );
    } else {
        eprintln!("\n{:<18} {:<14} {:<14} Pinned", "Tool", "Current", "Latest");
    }

    for e in &outdated {
        let name_trunc = &e.name[..e.name.len().min(17)];
        let cur_trunc = &e.installed[..e.installed.len().min(13)];
        let lat_trunc = &e.latest[..e.latest.len().min(13)];
        if e.pinned {
            if colored {
                eprintln!(
                    "{name_trunc:<18} {cur_trunc:<14} {lat_trunc:<14} {}",
                    style("~").dim()
                );
            } else {
                eprintln!("{name_trunc:<18} {cur_trunc:<14} {lat_trunc:<14} ~");
            }
        } else {
            eprintln!("{name_trunc:<18} {cur_trunc:<14} {lat_trunc:<14}");
        }
    }

    eprintln!("\nRun 'dot upgrade' to upgrade all (pinned skipped unless --force).\n");
    Ok(())
}
